### PR TITLE
Passing a variable by reference to a function and method has side effects

### PR DESCRIPTION
**Pull request by ryium**
- fixes https://github.com/phpstan/phpstan/issues/8084

### CODE REVIEW

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]
Could you please test and fix other similar functions that might have the same issue? array_push, array_unshift, array_pop, array_splice

**Comment by ondrejmirtes** [comment]
And maybe it'd be a better to idea to do this in `public function hasSideEffects(): TrinaryLogic` in various FunctionReflection and MethodReflection implementations - to return `TrinaryLogic::createYes()` if the function accepts parameters as references.

File: tests/PHPStan/Analyser/data/bug-8084.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+namespace Bug8084;
+
+use function array_shift;
+use function PHPStan\Testing\assertType;
+
+class Bug8084
+{
+	public function run(array $arr): void
+	{
+		/** @var array{a: 0, b?: 1} $arr */
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/bug-8084.php]: Hi, please put this type into `@param` above the method, here it's ambiguous what it means.

Also currently this example https://phpstan.org/r/de642079-e5b0-492a-9782-d78c324f70a7 shows `Expression on left side of ?? is not nullable.` so we'll need another test in NullCoalesceRuleTest to make sure this message goes away.

Thank you!

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: tests/PHPStan/Analyser/data/bug-8084.php (hunk 1 above)
Comment (ryium) [reply tests/PHPStan/Analyser/data/bug-8084.php]: Hi, thank you for your review.
Added a test, but included a merge commit to follow the latest branch.  
Should I recreate?

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Brought it over the finish line :) Thank you.

**Comment by ondrejmirtes** [comment]
Sorry, I had to revert this, it led to some false positives.

**Pull request merged**